File: Java/Design-Patterns/Abstract-Factory-Pattern/src/pizza/Pizza.java
package pizza;

import factory.PizzaIngredientFactory;
import pizza.ingredients.cheese.Cheese;
import pizza.ingredients.dough.Dough;
import pizza.ingredients.pepperoni.Pepperoni;

public abstract class Pizza {
    String name;
    PizzaIngredientFactory factory;
    Cheese cheese;
    Dough dough;
    Pepperoni pepperoni;

    public Pizza(String name, PizzaIngredientFactory factory) {
        this.name = name;
        this.factory = factory;
    }

    public abstract void prepare();

    public void bake() {
        System.out.println("Baking for 25 minutes at 350");
    }

    public void cut() {
        System.out.println("Cutting into triangle slices");
    }

    public void box() {
        System.out.println("Putting into a box");
    }

    public String getName() {
        return name;
    }
}

File: Scala/TourOfScala/Objects/src/Pizza.scala
sealed abstract class Pizza

class CheesePizza extends Pizza

class PepperoniPizza extends Pizza

object Pizza {

  def createPizza(pizzaType: String): Pizza = {
    var pizza: Pizza = null

    pizzaType match {
      case "cheese" => pizza = new CheesePizza
      case "pepperoni" => pizza = new PepperoniPizza
      case _ => pizza = new CheesePizza
    }

    pizza
  }

}
